### PR TITLE
LPS-134690 Categories section to be ordered into vocabularies

### PR DESCRIPTION
**Pull request by beltranrengifo**
# Motivation
The Content Dashboard Info Panel should organise the categories by vocabularies
[Jira Story](https://issues.liferay.com/browse/LPS-134690)

# Proposed Solution
Retrieve the vocabulary for each category, avoiding duplicates, and return an object to the frontend.
Also, the vocabulary group, and vocabulary visibility are retrieved.
The frontend component has been adapted.

# How to test and verify
- Create some vocabularies and categories
- Assign them to a document or any other asset (blog entry, web content article...)
- Visit the Content Dashboard
- Click in info icon of this asset, appears when the table row is hovered

Expected: The Categorisation section in the Info Panel should render the categories organised by vocabularies

image

### CODE REVIEW

**Comment by cgoncas** [comment]
Hey @beltranrengifo and @YurenaC 

I have needed to change some tests in https://github.com/liferay-tango/liferay-portal/pull/2102, so probably we will have a conflict with this one :(

Thanks!!

**Comment by YurenaC** [comment]
> Hey @beltranrengifo and @YurenaC
> 
> I have needed to change some tests in #2102, so probably we will have a conflict with this one :(
> 
> Thanks!!

No prob @cgoncas , let us know when your changes are ready and we'll compare, thanks!

File: modules/apps/content-dashboard/content-dashboard-test/src/testIntegration/java/com/liferay/content/dashboard/web/internal/portlet/action/test/GetContentDashboardItemInfoMVCResourceCommandTest.java [hunk 1]
```diff
@@ -75,16 +82,30 @@ public void setUp() throws Exception {
...
+
+		AssetVocabulary assetVocabulary = AssetTestUtil.addVocabulary(
+			_group.getGroupId());
+
+		AssetCategory assetCategory1 = AssetTestUtil.addCategory(
+			_group.getGroupId(), assetVocabulary.getVocabularyId());
+		AssetCategory assetCategory2 = AssetTestUtil.addCategory(
+			_group.getGroupId(), assetVocabulary.getVocabularyId());
+
+		JournalArticle journalArticle = _addJournalArticle(
```
Comment (cgoncas) [inline modules/apps/content-dashboard/content-dashboard-test/src/testIntegration/java/com/liferay/content/dashboard/web/internal/portlet/action/test/GetContentDashboardItemInfoMVCResourceCommandTest.java:94]: Hey @beltranrengifo and @YurenaC !!!

It is not a good practice to couple our test (of the content dashboard api) to the journal entity

Comment (cgoncas) [reply modules/apps/content-dashboard/content-dashboard-test/src/testIntegration/java/com/liferay/content/dashboard/web/internal/portlet/action/test/GetContentDashboardItemInfoMVCResourceCommandTest.java:94]: or the file entity

File: modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java [hunk 2]
```diff
@@ -182,18 +192,32 @@ protected void doServeResource(
 		}
 	}
 
-	private JSONArray _getAssetCategoriesJSONArray(
-		ContentDashboardItem contentDashboardItem, Locale locale) {
+	private HashMap<String, Object> _buildVocabularyDetails(
```
Comment (cgoncas) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java:195]: We should never return the implementation of a Map, better if we use the interface. (You can check the differences betwen interfacex Map and the implementation HasMap in https://www.javatpoint.com/map-and-hashmap-in-java#:~:text=The%20Map%20is%20an%20interface%2C%20and%20HashMap%20is%20a%20class,by%20using%20its%20implementing%20classes.&text=Whereas%20HashMap%20implements%20Map%20interface,the%20Map%20and%20HashMap%20objects.)

File: modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java [hunk 3]
```diff
@@ -182,18 +192,32 @@ protected void doServeResource(
...
+			Locale locale, AssetCategory assetCategory,
+			AssetVocabulary currentVocabulary)
```
Comment (cgoncas) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java:197]: As a Liferay style rule, we usually call the variable currentVocabulary with the same name that its type, so it should be renamed to assetVocabulary.

File: modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java [hunk 4]
```diff
@@ -182,18 +192,32 @@ protected void doServeResource(
...
+		throws PortalException {
 
-		List<AssetCategory> assetCategories =
-			contentDashboardItem.getAssetCategories();
+		return HashMapBuilder.<String, Object>put(
+			"categories",
+			() -> {
+				ArrayList<String> categoriesArray = new ArrayList<>();
```
Comment (cgoncas) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java:203]: The type of the variable should be the interface List, instead the implementation class ArrayList.

Following Liferay style book, its name should be categories.

File: modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java [hunk 5]
```diff
@@ -182,18 +192,32 @@ protected void doServeResource(
...
 
-		Stream<AssetCategory> stream = assetCategories.stream();
+				categoriesArray.add(assetCategory.getTitle(locale));
```
Comment (cgoncas) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java:205]: Why do we need to use an array, if we only add one value?

File: modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java [hunk 6]
```diff
@@ -182,18 +192,32 @@ protected void doServeResource(
...
-		return JSONUtil.putAll(
-			stream.map(
-				assetCategory -> assetCategory.getTitle(locale)
-			).toArray());
+				return categoriesArray;
+			}
+		).put(
+			"groupName",
+			ContentDashboardGroupUtil.getGroupName(
+				_groupLocalService.getGroup(currentVocabulary.getGroupId()),
```
Comment (cgoncas) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java:212]: When we call the method getGroup, a portal exception can be thrown, it is not better to use fetchGroup in this situation?

File: modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java [hunk 7]
```diff
@@ -209,6 +233,51 @@ private JSONArray _getAssetTagsJSONArray(
...
+
+		Stream<AssetCategory> stream = assetCategories.stream();
+
+		HashMap<Long, Map<String, Object>> assetVocabularies = new HashMap<>();
+
+		stream.forEach(
+			assetCategory -> {
+				long vocabularyId = assetCategory.getVocabularyId();
+
+				if (Validator.isNull(assetVocabularies.get(vocabularyId))) {
```
Comment (cgoncas) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java:250]: Better if we use Optional interface instead https://www.baeldung.com/java-optional

File: modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java [hunk 8]
```diff
@@ -209,6 +233,51 @@ private JSONArray _getAssetTagsJSONArray(
...
+					AssetVocabulary currentVocabulary =
+						_assetVocabularyLocalService.fetchAssetVocabulary(
+							assetCategory.getVocabularyId());
+
+					try {
+						HashMap<String, Object> vocabularyDetails =
```
Comment (cgoncas) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java:256]: As this variable is only used once, we don't need to define it.

**Review by cgoncas — CHANGES_REQUESTED** [review]
Hey @beltranrengifo and @YurenaC !!!

I have added some minor changes, let me know if you need any clarification.

Thanks!!!

File: modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java (hunk 5 above)
Comment (beltranrengifo) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java:205]: We are calling `_build Vocabulary Details` once for each category if it's vocabulary hasn't been seen before.
We add the rest of the categories [here](https://github.com/liferay-tango/liferay-portal/blob/e5f83e7872e3132714349532687d0b9c59b0290a/modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java#L274)

**Comment by beltranrengifo** [comment]
Hi @cgoncas thanks for the comments! 🎉 
Everything is done & pushed, except the first one:

> It is not a good practice to couple our test (of the content dashboard api) to the journal entity

I don't get how to test the output if we do not create/use any type of file in the test 🤔

File: modules/apps/content-dashboard/content-dashboard-test/src/testIntegration/java/com/liferay/content/dashboard/web/internal/portlet/action/test/GetContentDashboardItemInfoMVCResourceCommandTest.java (hunk 1 above)
Comment (YurenaC) [reply modules/apps/content-dashboard/content-dashboard-test/src/testIntegration/java/com/liferay/content/dashboard/web/internal/portlet/action/test/GetContentDashboardItemInfoMVCResourceCommandTest.java:138]: How do you suggest to create the content instead? 
The first thing that comes to my mind in this case, is create a DB script, like a set up that creates all the data needed, but it is not a common practice in the portal either.

Happy to hear any ideas :D

**Comment by beltranrengifo** [comment]
ci:test:relevant

**Comment by beltranrengifo** [comment]
ci:test:relevant

File: modules/apps/content-dashboard/content-dashboard-test/src/testIntegration/java/com/liferay/content/dashboard/web/internal/portlet/action/test/GetContentDashboardItemInfoMVCResourceCommandTest.java (hunk 1 above)
Comment (cgoncas) [reply modules/apps/content-dashboard/content-dashboard-test/src/testIntegration/java/com/liferay/content/dashboard/web/internal/portlet/action/test/GetContentDashboardItemInfoMVCResourceCommandTest.java:138]: We should follow the same approach that we do in com.liferay.analytics.reports.test.analytics.reports.info.item.MockObjectAnalyticsReportsInfoItem. So we create an entity for the tests.

So we could decouple the API for any implementation. This way,  the changes in Journal won 't affect the ContentDashboardAPI

File: modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java [hunk 9]
```diff
@@ -209,6 +223,48 @@ private JSONArray _getAssetTagsJSONArray(
...
+		Map<Long, Map<String, Object>> assetVocabularies = new HashMap<>();
+
+		stream.forEach(
+			assetCategory -> {
+				long vocabularyId = assetCategory.getVocabularyId();
+
+				Optional<Map<String, Object>> vocabularyOptional =
+					Optional.ofNullable(assetVocabularies.get(vocabularyId));
+
+				if (vocabularyOptional.isPresent()) {
```
Comment (cgoncas) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java:243]: The idea of use Optional in this case is so we can have objects that are created just one time, so they can be inmutables.

You could implement it as something like:

Map<Long, Map<String, Object>> assetVocabularies = stream.map ...

File: modules/apps/content-dashboard/content-dashboard-test/src/testIntegration/java/com/liferay/content/dashboard/web/internal/portlet/action/test/GetContentDashboardItemInfoMVCResourceCommandTest.java (hunk 1 above)
Comment (cgoncas) [reply modules/apps/content-dashboard/content-dashboard-test/src/testIntegration/java/com/liferay/content/dashboard/web/internal/portlet/action/test/GetContentDashboardItemInfoMVCResourceCommandTest.java:138]: You can see a full example of how we use it in com.liferay.analytics.reports.web.internal.item.action.provider.test.AnalyticsReportsContentDashboardItemActionProviderTest.

File: modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java (hunk 9 above)
Comment (cgoncas) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java:243]: For example something like

Map<Long, Map<String, List<String>> assetVocabularies = new HashMap<>();

		stream.forEach(
			assetCategory -> {
				assetVocabularies.computeIfAbsent(
					vocabularyId -> {
						AssertVocabulary assetVocabulary =
							_assetVocabularyLocalService.fetchAssetVocabulary(
								vocabularyId);

						return_HashMapBuilder.<String, Object>put(
							"categories", new ArrayList<>()
						).put(
							"groupName",
							ContentDashboardGroupUtil.getGroupName(
								_groupLocalService.fetchGroup(
									assetVocabulary.getGroupId()),
								locale)
						).put(
							"isPublic",
							assetVocabulary.getVisibilityType() ==
							AssetVocabularyConstants.VISIBILITY_TYPE_PUBLIC
						).put(
							"vocabularyName",
							assetVocabulary.getTitleCurrentValue()
						).build();
					}
				);


			}).get();

Comment (cgoncas) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java:243]: A good example of how we can use compute if absent can be found in https://www.baeldung.com/java-map-computeifabsent

Comment (beltranrengifo) [reply modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java:243]: Ok, thxs, gonna try.
The `assetVocabularies` HashMap cannot be declared and assigned to the stream, right?

**Comment by cgoncas** [comment]
> Hi @cgoncas thanks for the comments! 🎉 Everything is done & pushed, except the first one:
> 
> > It is not a good practice to couple our test (of the content dashboard api) to the journal entity
> 
> I don't get how to test the output if we do not create/use any type of item in the test 🤔

We could follow an approach like the one we follow in com.liferay.analytics.reports.test.util.MockContextUtil or we can delete this class, and keep going with the Unit test defined in GetContentDashboardItemInfoMVCResourceCommandTest.java.

File: modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java [hunk 10]
```diff
@@ -210,6 +210,34 @@ protected void doServeResource(
 		).build();
 	}
 
+	private Collector<AssetCategory, ?, Map<Long, Map<String, Object>>>
+		_collectVocabularies(Locale locale) {
```
Comment (cgoncas) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java:214]: I think that if we follow Liferay name guides, this should be called _getCollector(

File: modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java [hunk 11]
```diff
@@ -210,6 +210,34 @@ protected void doServeResource(
...
+
+		return Collector.of(
```
Comment (cgoncas) [inline modules/apps/content-dashboard/content-dashboard-web/src/main/java/com/liferay/content/dashboard/web/internal/portlet/action/GetContentDashboardItemInfoMVCResourceCommand.java:216]: This is the most beautiful thing I have ever see.

**Comment by beltranrengifo** [comment]
> ollow in com.liferay.analytics.reports.test.util.MockContextUtil or we can delete this class, and keep going with the Unit test d

You mean delete the entire `GetContentDashboardItemInfoMVCResourceCommandTest.java` class?
Are the units already done?
Many questions I have, sorry 😄

**Comment by claraizquierdo** [comment]
Hey, nice work, everything look very good. Just two minor comments:

1. The site name is always shown, even for Global vocabularies. In the Figma I don't see the site name when the vocabulary is Global. Should be always visible or not?
Captura de pantalla 2022-01-25 a las 13 53 11

2. The line height of the text "PUBLIC CATEGORIES" and "INTERNAL CATEGORIES" does not match with the Figma. I don't know which one is ok.

**Comment by beltranrengifo** [comment]
> Hey, nice work, everything look very good. Just two minor comments:
> 
> 1. The site name is always shown, even for Global vocabularies. In the Figma I don't see the site name when the vocabulary is Global. Should be always visible or not?
>    Captura de pantalla 2022-01-25 a las 13 53 11
> 2. The line height of the text "PUBLIC CATEGORIES" and "INTERNAL CATEGORIES" does not match with the Figma. I don't know which one is ok.

Hi @claraizquierdo !
Thanxs for pointing on that.
- I'm not sure neither about rendering the global/non-global vocabulary names. Figma only renders non-globals, but I really do not know if it is the desired behaviour. I'll ask
- I'll check that line height 👍

**Comment by beltranrengifo** [comment]
ci:test:sf

**Comment by beltranrengifo** [comment]
ci:test:relevant

**Review by claraizquierdo — APPROVED** [review]

**Comment by beltranrengifo** [comment]
ci:test:tango

**Comment by beltranrengifo** [comment]
ci:forward

**Pull request closed without merging**